### ISSUE TITLE
OIDC authentication multiple providers tests

### ISSUE THREAD

**Issue by tleyden** [couchbase/sync_gateway#1882]
Factored out of https://github.com/couchbase/sync_gateway/issues/1774
- [ ] Define multiple providers targeting the same issuer in the config.  Get an ID token for one of the providers, and verify that you can use it to authenticate successfully (ID token authentication relies on looking up the provider based on the issuer AND the audience (client_id) defined in the token).  Requires fix for  https://github.com/couchbase/sync_gateway/issues/1878
- [ ] Define multiple providers targeting the same issuer in the config.  Authenticate and trigger user creation using one of the providers.  Authenticate as the same user via the other provider, and make sure it authenticates as the same user.  Requires fix for https://github.com/couchbase/sync_gateway/issues/1876

**Comment by tleyden** [comment]
If you're setting up an app for google signin, google has separate client ids for separate platforms (android, ios).  Both use same registered app with google, multiple client ids.  Down to the way that google has native integration.  

Example config:

```
"providers": {
            "test2": {
               "issuer":"http://localhost:4984/todos/_oidc_testing",
               "client_id":"sync_gateway_2",
               "validation_key":"R75hfd9lasdwertwerutecw8",
               "callback_url":"http://localhost:4984/todos/_oidc_callback",
              },
            "test": {
               "issuer":"http://localhost:4984/todos/_oidc_testing",
               "client_id":"sync_gateway",
               "validation_key":"R75hfd9lasdwertwerutecw8",
               "callback_url":"http://localhost:4984/todos/_oidc_callback",
               "register":true
           },
```
- Get token for `test`
- If it used test2, token wouldn't work (make sure happy path works)

**Comment by tleyden** [comment]
2nd one: 
- Two providers
- Both client_id's are valid

```
"providers": {
            "testprovider2": {
               "issuer":"http://localhost:4984/todos/_oidc_testing",
               "client_id":"sync_gateway_2",
               "validation_key":"R75hfd9lasdwertwerutecw8",
               "callback_url":"http://localhost:4984/todos/_oidc_callback",
               "register":true
              },
            "testprovider": {
               "issuer":"http://localhost:4984/todos/_oidc_testing",
               "client_id":"sync_gateway",
               "validation_key":"R75hfd9lasdwertwerutecw8",
               "callback_url":"http://localhost:4984/todos/_oidc_callback",
               "register":true
           },
```

Login using `testprovider` (for example, with an ios device), but then use the id token on `testprovider2` (for example, with an android device) and the id token should still work

**Comment by ajres** [comment]
To support this in the test provider I think we could just change #274 in handler.go to take the username instead of the 'testProviderAud' constant:

`provider := context.Options.OIDCOptions.Providers.GetProviderForIssuer(issuerUrlForDB(h, context.Name), testProviderAud)`

to

`provider := context.Options.OIDCOptions.Providers.GetProviderForIssuer(issuerUrlForDB(h, context.Name), username)`

I don’t think this opens up any security holes.

**Comment by tleyden** [comment]
@adamcfraser regarding:

> Define multiple providers targeting the same issuer in the config. Get an ID token for one of the providers, and verify that you can use it to authenticate successfully (ID token authentication relies on looking up the provider based on the issuer AND the audience (client_id) defined in the token). Requires fix for #1878

I tried testing with the config mentioned in https://github.com/couchbase/sync_gateway/issues/1882#issuecomment-227823487, but when I added tracing I could see that it's not really a valid test because it is looping over the keys of an unordered map, and it hits `test` before `test2` anyway.

These tests are feeling like they might be better done and more maintainable as unit tests.  Going to close this ticket for now.

**Comment by tleyden** [comment]
TODO: re-add config entry, depend on hashmap shuffling to test
